src/schnetpack/nn/embed.py

```diff
@@ -0,0 +1,278 @@
+
+        # Scale pairwise distance vectors with pairwise mask
+        r_ij = safe_scale(r_ij, scale=pair_mask[:, None])
+
+        # Calculate pairwise distances
+        d_ij = safe_scale(np.linalg.norm(r_ij, axis=-1), scale=pair_mask)  # shape : (n_pairs)
+
+        # Gaussian basis expansion of distances
+        rbf_ij = safe_scale(self.rbf_fn(d_ij[:, None]), scale=pair_mask[:, None])  # shape: (n_pairs,K)
+        phi_r_cut = safe_scale(self.cut_fn(d_ij), scale=pair_mask)  # shape: (n_pairs)
+
+        # Normalized distance vectors
+        unit_r_ij = safe_mask(mask=d_ij[:, None] != 0,
+                              operand=r_ij,
+                              fn=lambda y: y / d_ij[:, None],
+                              placeholder=0
+                              )  # shape: (n_pairs, 3)
+        unit_r_ij = safe_scale(unit_r_ij, scale=pair_mask[:, None])  # shape: (n_pairs, 3)
+
+        # Spherical harmonics
+        sph_harms_ij = []
+        for sph_fn in self.sph_fns:
+            sph_ij = safe_scale(sph_fn(unit_r_ij), scale=pair_mask[:, None])  # shape: (n_pairs,2l+1)
+            sph_harms_ij += [sph_ij]  # len: |L| / shape: (n_pairs,2l+1)
+
+        sph_harms_ij = np.concatenate(sph_harms_ij, axis=-1) if len(self.degrees) > 0 else None
+        # shape: (n_pairs,m_tot)
+
+        geometric_data = {'R': R,
+                          'r_ij': r_ij,
+                          'unit_r_ij': unit_r_ij,
+                          'd_ij': d_ij,
+                          'rbf_ij': rbf_ij,
+                          'phi_r_cut': phi_r_cut,
+                          'sph_ij': sph_harms_ij,
+                          }
+
+        # Spherical harmonic coordinates (SPHCs)
+        if self.sphc:
+            z = inputs[self.atomic_type_key]
+            point_mask = inputs['point_mask']
+            if self.sphc_normalization is None:
+                # Initialize SPHCs to zero
+                geometric_data.update(_init_sphc_zeros(z=z,
+                                                       sph_ij=sph_harms_ij,
+                                                       phi_r_cut=phi_r_cut,
+                                                       idx_i=idx_i,
+                                                       point_mask=point_mask,
+                                                       mp_normalization=self._lambda)
+                                      )
+            else:
+                # Initialize SPHCs with a neighborhood dependent embedding
+                geometric_data.update(_init_sphc(z=z,
+                                                 sph_ij=sph_harms_ij,
+                                                 phi_r_cut=phi_r_cut,
+                                                 idx_i=idx_i,
+                                                 point_mask=point_mask,
+                                                 mp_normalization=self._lambda)
+                                      )
+
```

Comment:
I would recommend to exclude all SPHC parts in a seperate module. Please make sure to use torch instead of numpy